agent/xds/server.go

```diff
@@ -398,7 +398,7 @@ func (t *xDSType) SendIfNew(cfgSnap *proxycfg.ConfigSnapshot, version uint64, no
 	// for the EDS response. This is fixed though by ensuring we send an explicit
 	// empty LoadAssignment resource for the cluster rather than allowing junky
 	// empty resources.
-	if len(resources) == 0 {
+	if len(resources) == 0 && !allowEmpty {
```

Comment:
hm, this approach seems like it changes a lot of code. but does it have to?
could this line just be, `if len(resources) == 0 && cfgSnap.Kind != structs.ServiceKindMeshGateway` with the comment you had above noting that only mesh gateways can `inform CDS of no clusters`?

Reply:
I see where you are coming from but allow me to clarify something first. The protocol for chatting back and forth with envoy is called `xDS` because the `x` is like a placeholder for `C(luster)`, L(istener)`, R(oute)`, E(ndpoint)`, etc. In the context of this issue I'm only loosening the restriction for `CDS` so that we can legitimately say that having "no clusters" (or in this case "no connect enabled services registered") is acceptable.

This `agent/xds/server.go` file is generic plumbing for updating all kinds of `x` things, so is common for `CDS`, `EDS`, `RDS`, and `LDS`. The suggestion you are making would allow for the code to send back "no listeners", "no routes", and "no endpoints" in addition to the desired "no clusters" result.

Reply:
oh I see, thanks for the context. how about this:
```
allowEmpty := t.typeURL == ClusterType && cfgSnap.Kind != structs.ServiceKindMeshGateway
```

Reply:
I did something in the same spirit as your suggestion.